### ISSUE TITLE
Document best practices or tips and tricks for updating or iterating over confd-managed files

### ISSUE THREAD

**Issue by emetsger** [Islandora-Devops/isle-dc#119]
[confd](http://www.confd.io/) is used by ISLE 8 Docker containers to create relevant configuration files at runtime from [template resources](https://github.com/kelseyhightower/confd/blob/master/docs/template-resources.md) and template files.  confd provides a uniform approach for the interpolation of environment variables in configuration files used by ISLE.

For an example, see the `nginx.conf` [template](https://github.com/Islandora-Devops/isle-buildkit/blob/main/nginx/rootfs/etc/confd/templates/nginx.conf.tmpl) and corresponding [template resource](https://github.com/Islandora-Devops/isle-buildkit/blob/main/nginx/rootfs/etc/confd/conf.d/nginx.conf.toml).

At container boot prior to any other services starting, confd is [invoked](https://github.com/Islandora-Devops/isle-buildkit/blob/main/base/rootfs/etc/cont-init.d/01-confd-render-templates.sh) to produce the configuration files using the environment variables in the template and replacing them with their runtime value.

Not only does confd run at container boot time, it runs [_all_ the time](https://github.com/Islandora-Devops/isle-buildkit/blob/main/base/rootfs/etc/services.d/confd/run), regenerating config files from templates the environment according to the value of [`CONFD_POLLING_INTERVAL`](https://github.com/Islandora-Devops/isle-buildkit/blob/main/base/rootfs/usr/local/bin/confd-render-templates.sh#L12).

This is where things get a little tricky for those new to ISLE, specifically those unfamiliar with confd (or S6 for that matter).  A developer familiar with Docker and Nginx (or any of the other relevant services under confd's templating control) may wish to modify the files at runtime (e.g. `nginx.conf`, `php.ini`) using bind mounts or direct editing of the files within the container.  They will be surprised when their changes are seemingly overwritten at random times.  Eventually, they'll dig into the containers and uncover the use of confd, and then research and understand its use in order to become proficient at testing or iterating over their customizations to configuration files.

This issue simply serves to flag this particular issue, which can probably be resolved with documentation.

It does raise a question, however: how many people are relying on the continuous regeneration of configuration files?  And would it make sense to make continuous regeneration optional, and have it disabled by default?  Granted, any customizations to confd managed files will need to be placed into their corresponding template; there is no getting around that if the "ISLE way" is being followed.   But it may make ISLE just a little more friendly to the newcomer.  Regardless, documenting approaches for updating or iterating over confd managed files will go a long way.

I think a couple of options were mentioned at the last meeting:
* replace the template (e.g. the [nginx.conf.tmpl](https://github.com/Islandora-Devops/isle-buildkit/blob/main/nginx/rootfs/etc/confd/templates/nginx.conf.tmpl) with a concrete configuration configuration file
  * (this assumes the user knows where to look, which would be provided by documentation)
* optionally disable confd regeneration while iterating
* set `CONFD_POLLING_INTERVAL` to some ridiculously large number

**Comment by nigelgbanks** [comment]
Documenting sounds like a good idea, I just have few wrenches to add about about disabling the `confd` as service. Unfortunately with `s6` there is no way to have a disabled service as far as I could find. So it's keep or remove it as a service. I'm fine either way as I can add the service back in where needed for my use case. 

One reason to have `confd` act as a service would be to reload nginx configuration at runtime (though this isn't implemented in the current nginx image). An example would be using different backend other than environment variables driving `confd` and periodically updating that back end with a new cert, at which point  `nginx` configuration would get regenerated and the service reloaded without the container going down so the site has zero down time. https://github.com/kelseyhightower/confd/blob/master/docs/template-resources.md#notes

**Comment by emetsger** [comment]
Yeah, I don't mean to imply that `confd` should be removed as a service.  However, you could influence its behavior based on an environment variable.

> * optionally disable confd regeneration while iterating

This might be enabled by supporting an env var like `CONFD_NOOP`, which is used to alter the behavior of the service.  Below is an example; the logic is a bit backward, but nevertheless communicates the general idea.

```diff
diff --git a/base/rootfs/etc/services.d/confd/run b/base/rootfs/etc/services.d/confd/run
index eedca16..7608e15 100644
--- a/base/rootfs/etc/services.d/confd/run
+++ b/base/rootfs/etc/services.d/confd/run
@@ -1,3 +1,9 @@
 #!/usr/bin/with-contenv bash
-set -e 
-confd-render-templates.sh --continuous
+set -vxe
+
+if [ -n "${CONFD_NOOP}" ] && [ "false" != "${CONFD_NOOP}" ]
+then
+  confd-render-templates.sh
+else
+  confd-render-templates.sh --continuous
+fi
```

**Comment by nigelgbanks** [comment]
That alone may not do it, there is a `s6`  [finish script](https://github.com/just-containers/s6-overlay#writing-an-optional-finish-script), that is [setup to currently shutdown the container](https://github.com/Islandora-Devops/isle-buildkit/blob/main/base/rootfs/etc/services.d/confd/finish) if the [run script](https://github.com/Islandora-Devops/isle-buildkit/blob/main/base/rootfs/etc/services.d/confd/run) exits. If no `finish` script is provided it `s6` will just re-run the `run` script again, there may be a way of having the `finish` script do nothing and then exit leaving the container running. I haven't tested and it's doesn't state anything to the affect in the `s6` README. 

If the `finish` script can be worked around then what you suggest could definitely work. 

Small thing to add. We do not need to render the script once since that is already handled [here](https://github.com/Islandora-Devops/isle-buildkit/blob/main/base/rootfs/etc/cont-init.d/01-confd-render-templates.sh).

```diff
diff --git a/base/rootfs/etc/services.d/confd/run b/base/rootfs/etc/services.d/confd/run
index eedca16..7608e15 100644
--- a/base/rootfs/etc/services.d/confd/run
+++ b/base/rootfs/etc/services.d/confd/run
@@ -1,3 +1,9 @@
 #!/usr/bin/with-contenv bash
-set -e 
-confd-render-templates.sh --continuous
+set -e
+
+if [[ ! -z "${CONFD_ENABLE_SERVICE}" ]]; then
+  confd-render-templates.sh --continuous
+fi
```

**Comment by nigelgbanks** [comment]
I suppose we could always `sleep` to get around the `finish` problem, though that does feel oddly icky. It does require a fixed length of time `2^31-1=2147483647` would be the max number I think it takes, which is like 70 years or some such.

`sleep 2147483647`

**Comment by emetsger** [comment]
> Small thing to add. We do not need to render the script once since that is already handled here.

👌 Right! 

> That alone may not do it, there is a s6 finish script, that is setup to currently shutdown the container if the run script exits.

I thought I tested that diff at one point (the diff that I posted originally) and it seemed to work with no other modifications, but I may be mistaken.  (It was applied against an older version of isle-buildkit then updated to work with the new-ish `confd-render-templates.sh` script, so who knows)

In any case, this discussion is super helpful, thanks @nigelgbanks!

From my perspective, it sounds like there are options for "disabling" or "neutering" the ongoing rendering of templates, _if_ folks agree that it would be a helpful way to address the usecase originally documented in this issue (iterating over edits of confd-managed files at runtime).  

For example, documenting an approach that has one (un)set an ENV var like `CONFD_ENABLE_SERVICE` may be a lot simpler than attempting to document the location of each and every template file and asking folks to replace the template with their concrete copy.  An alternate, maybe just as easy approach, is to set `CONFD_POLLING_INTERVAL` to `2147483647`.  Which ever is easiest to document and support would have my vote.

**Comment by nigelgbanks** [comment]
Sounds like a good idea for me, I would also put that it should be disabled by default as the use case for running `confd` as a service isn't really fulfilled at the moment.

**Comment by emetsger** [comment]
> I would also put that it should be disabled by default as the use case for running confd as a service isn't really fulfilled at the moment.

Agreed!
